From 2eb9063b213e85bdf1e1bc764fc040febfed910f Mon Sep 17 00:00:00 2001
From: Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
Date: Tue, 25 Jun 2024 11:24:21 +0000
Subject: [PATCH] [MSE][GStreamer] block data flow until source pad is expoded

There could be a race between main thread (linking typefind element, see
gsturisourcebin.c:setup_typefind) and streaming thread (pushing
data). webKitMediaSrcLoop does wait for source pad to linked. However,
there is no guaranty that peer pad is already active at that
point. Which may lead to data loss while peer the pad is being
activated.

This change block downstream flow during pad addition, allowing
urisourcebin to link/activate typefind sink pad.
---
 .../gstreamer/mse/WebKitMediaSourceGStreamer.cpp      | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp b/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp
index 4f3617dd20736..1731f3237095a 100644
--- a/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp
@@ -41,6 +41,7 @@
 #include <wtf/text/AtomString.h>
 #include <wtf/text/AtomStringHash.h>
 #include <wtf/text/CString.h>
+#include <wtf/Scope.h>
 
 using namespace WTF;
 using namespace WebCore;
@@ -327,6 +328,16 @@ void webKitMediaSrcEmitStreams(WebKitMediaSrc* source, const Vector<RefPtr<Media
     gst_element_post_message(GST_ELEMENT(source), gst_message_new_stream_collection(GST_OBJECT(source), source->priv->collection.get()));
 
     for (const RefPtr<Stream>& stream: source->priv->streams.values()) {
+        // Block data flow until pad is exposed
+        gulong blockId = gst_pad_add_probe (
+            GST_PAD(stream->pad.get()), GST_PAD_PROBE_TYPE_BLOCK_DOWNSTREAM,
+            [](GstPad*, GstPadProbeInfo*,gpointer) -> GstPadProbeReturn {
+                return GST_PAD_PROBE_OK;
+            }, nullptr, nullptr);
+        auto blockCleanup = WTF::makeScopeExit([&] {
+            gst_pad_remove_probe(GST_PAD(stream->pad.get()), blockId);
+        });
+
         if (!webkitGstCheckVersion(1, 20, 6)) {
             // Workaround: gst_element_add_pad() should already call gst_pad_set_active() if the element is PAUSED or
             // PLAYING. Unfortunately, as of GStreamer 1.18.2 it does so with the element lock taken, causing a deadlock
